examples/reporting/parallel_report_download.py

```diff
@@ -0,0 +1,159 @@
+#!/usr/bin/env python
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Shows how to download a set of reports from a list of accounts in parallel.
+
+If you need to obtain a list of accounts, please see the
+account_management/get_account_hierarchy.py or
+account_management/list_accessible_customers.py examples.
+"""
+
+import argparse
+from itertools import product
+import multiprocessing
+import sys
+import time
+
+from google.ads.google_ads.client import GoogleAdsClient
+from google.ads.google_ads.errors import GoogleAdsException
+
+# Maximum number of processes to spawn
+MAX_PROCESSES = multiprocessing.cpu_count()
+# Timeout between retries in seconds
+BACKOFF_FACTOR = 5
+# Maximum number of retries for 500 errors
+MAX_RETRIES = 5
+
+
+def main(client, customer_ids):
+    """The main method that creates all necessary entities for the example.
+
+    Args:
+        client: an initialized GoogleAdsClient instance.
+        customer_ids: an array of client customer IDs.
+    """
+
+    # Define the GAQL query strings to run for each customer ID.
+    campaign_query = """
+        SELECT campaign.id, metrics.impressions, metrics.clicks
+        FROM campaign
+        WHERE segments.date DURING LAST_30_DAYS"""
+    ad_group_query = """
+        SELECT campaign.id, ad_group.id, metrics.impressions, metrics.clicks
+        FROM ad_group
+        WHERE segments.date DURING LAST_30_DAYS"""
+
+    inputs = _generate_inputs(
+        client, customer_ids, [campaign_query, ad_group_query]
+    )
+    with multiprocessing.Pool(MAX_PROCESSES) as pool:
+        results = pool.starmap(_issue_search_request, inputs)
+        print(results)
+
+
+def _issue_search_request(client, customer_id, query):
+    """Issues a search request using streaming.
+
+    If a 500 error is received, retries until MAX_RETRIES is reached.
```

Comment:
@wihl @BenRKarl I haven't been able to find a way to determine the http status code from the GoogleAdsException but it definitely seemed like something that should be accessible 🤔 Any ideas or examples you can point me to?

Reply:
Should we even retry a 500 error? If it truly is an internal error, we should not retry

Reply:
I was basing that off of the adwords [example](https://github.com/googleads/googleads-python-lib/blob/master/examples/adwords/v201809/reporting/parallel_report_download.py) but actually the golden does not appear to be looking for specific error codes so maybe just basing retries on errors more generally is ok? Unless you want to specifically not retry on 500 errors (in my limited experience, these can be transient, so I'm not totally on board with not retrying on them)

Reply:
Since max retries is only 5, I think it is ok to retry on all errors including 500's. We should investigate why you are getting transient 500 errors. That's not a normal situation.

Reply:
That was when running the ft framework, I'll share the bug with you directly; maybe it's less of a concern in that situation.